codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -496,19 +530,52 @@ private void incrementPendingBytes(int numBytes) {
              */
             int write(int allowedBytes) {
                 int before = payload.size();
-                needFlush |= payload.write(Math.max(0, allowedBytes));
-                int writtenBytes = before - payload.size();
+                int writtenBytes = 0;
+                try {
+                    if (writing) {
+                        throw new IllegalStateException("write is not re-entrant");
+                    }
+                    // Write the portion of the frame.
+                    writing = true;
+                    needFlush |= payload.write(Math.max(0, allowedBytes));
+                    if (!cancelled && payload.size() == 0) {
+                        // This frame has been fully written, remove this frame
+                        // and notify the payload. Since we remove this frame
+                        // first, we're guaranteed that its error method will not
+                        // be called when we call cancel.
+                        pendingWriteQueue.remove();
+                        payload.writeComplete();
+                    }
+                } catch (Throwable e) {
+                    // Mark the state as cancelled, we'll clear the pending queue
+                    // via cancel() below.
+                    cancelled = true;
```

Comment:
@Scottmitch I discussed this with Nathan actually. We don't want to rethrow for `write()`, but we could do it for `writeComplete()`, but then decided to only rethrow if `error()` throws an Exception. I believe we thought that it wasn't worth the extra complexity it would add to the code (distinguishing between write() throwing and writeComplete() throwing). What's your thoughts on this?